Mathlib/Data/DigitExpansion/Defs.lean

```diff
@@ -0,0 +1,504 @@
+/-
+Copyright (c) 2022 Yakov Pechersky. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Yakov Pechersky
+-/
+import Mathlib.Order.Fin
+import Mathlib.Order.SuccPred.Basic
+import Mathlib.Tactic.Ring
+
+
+/-!
+# Defining reals without the use of rationals
+
+* [Defining reals without the use of rationals][debruijn1976]
+
+-/
+
+/-- A possibly infinite sequence of digits in `Fin (b + 1)`, which
+can represent a number system like the reals.
+It is constrained to not end in an infinite string of the top digit. -/
+structure DigitExpansion (Z : Type*) [LT Z] (b : ℕ) where
+  /-- The digit sequence, such that `b` is the base - 1. -/
+  toFun : Z → Fin (b + 1)
+  /-- The constraint that the sequence does not trail end infinitely at the top digit. -/
+  bounded' : ¬∃ i₀, ∀ i > i₀, toFun i = b
```

Comment:
When `b = 2`, this doesn't ban an infinite series of `1` at the top digits, right?

Reply:
I guess I'm not clear on whether this is trying to ban `...222222121.01` or `012.12122222222...`

Reply:
0.999999... is banned in base 10.

Reply:
I think then I'd argue that your digits are backwards; I'd expect digit `1` to be the tens digit, not the tenths digit (to match the order of Nat.digits)

Reply:
I was following de Bruijn's convention: 
![image](https://github.com/user-attachments/assets/3a7d3ef7-2bdf-4a69-8714-fe30671785b4)

Reply:
Even though he states that the sum uses `b^i`, later in the paragraph, he refers to "non-negative reals [...] a_i = 0 for all i < i_0", so that means left of the decimal point is negative.